File: agent/tools/AnalyzeTool.js
const fs = require('fs');
const BaseTool = require('./BaseTool');
const OpenAI = require('openai');

class AnalyzeTool extends BaseTool {
  constructor(agent) {
    super(agent);
    this.openai = new OpenAI({
      apiKey: process.env.OPENAI_API_KEY
    });
  }

  async execute() {
    this.log('📊 Analyzing files for hardcoded strings...');
    
    // Get files from agent's memory/state
    const state = this.agent.stateManager.getState();
    const files = state.context.filesToProcess || [];
    
    this.log(`  🧠 Retrieved files from memory: ${JSON.stringify(files)}`);
    this.log(`  🧠 Files type: ${typeof files}, length: ${files.length}`);
    
    if (files.length === 0) {
      this.log('  ❌ No files to analyze');
      return { filesAnalyzed: 0, stringsFound: 0, files: {} };
    }
    
    const analysisResults = {
      filesAnalyzed: 0,
      stringsFound: 0,
      files: {}
    };
    
    for (const filePath of files) {
      if (fs.existsSync(filePath)) {
        this.log(`  📄 Analyzing ${filePath}...`);
        const content = fs.readFileSync(filePath, 'utf8');
        this.log(`  📄 File content length: ${content.length} characters`);
        this.log(`  📄 File content preview: ${content.substring(0, 300)}...`);
        
        const strings = await this.findHardcodedStringsWithGPT(content, filePath);
        
        this.log(`  📊 Analysis complete for ${filePath}: ${strings.length} strings found`);
        
        if (strings.length > 0) {
          analysisResults.filesAnalyzed++;
          analysisResults.stringsFound += strings.length;
          analysisResults.files[filePath] = {
            strings: strings,
            count: strings.length,
            complexity: this.calculateComplexity(strings),
            needsUseTranslation: !content.includes('useTranslation'),
            hasI18nSetup: content.includes('next-i18next') || content.includes('react-i18next')
          };
          
          this.log(`  📝 Found ${strings.length} strings in ${filePath}`);
          this.log(`  📝 Strings: ${strings.map(s => s.text).join(', ')}`);
        } else {
          this.log(`  ⚠️ No strings found in ${filePath}`);
        }
      } else {
        this.log(`  ❌ File not found: ${filePath}`);
      }
    }
    
    // Store results in agent state and update phase
    this.agent.stateManager.updateContext({ 
      stringsFound: analysisResults.stringsFound,
      analysisResults: analysisResults 
    });
    
    this.agent.stateManager.updateState({
      phase: 'analyzing'
    });
    
    return analysisResults;
  }

  async findHardcodedStringsWithGPT(content, filePath) {
    try {
      this.log(`  🤖 Calling GPT for ${filePath}...`);
      
      const prompt = `Find ALL user-facing text in this React file. Return JSON with phrases array.

INCLUDE:
- Button text content
- Link text content  
- Label text content
- Alt text (alt="...")
- Placeholder text (placeholder="...")
- Aria labels (aria-label="...")
- Title attributes (title="...")
- Any visible text users see
- Error messages, notifications, tooltips
- Form labels and help text

EXCLUDE:
- HTML IDs, CSS classes, variable names
- Code logic, function names
- URLs, file paths, technical values
- Comments and console.log statements

Generate semantic keys like "button.save", "menu.my_books", "alt.logo", "placeholder.email"

Return only valid JSON: {"phrases": [{"text": "original", "key": "semantic.key"}]}

File: ${filePath}

\`\`\`jsx
${content}
\`\`\``;

      this.log(`  📝 Prompt length: ${prompt.length} characters`);
      
      const response = await this.openai.chat.completions.create({
        model: 'gpt-4o-mini',
        messages: [{ role: 'user', content: prompt }],
        temperature: 0.1,
        max_tokens: 2000
      });

      this.log(`  📥 GPT response received`);
      
      let responseText = response.choices[0].message.content;
      this.log(`  📄 Raw response length: ${responseText.length} characters`);
      this.log(`  📄 Raw response preview: ${responseText.substring(0, 200)}...`);
      
      // Extract JSON from markdown code blocks if present
      const jsonMatch = responseText.match(/```(?:json)?\s*(\{[\s\S]*?\})\s*```/);
      if (jsonMatch) {
        responseText = jsonMatch[1];
        this.log(`  🔍 Extracted JSON from markdown`);
      } else {
        this.log(`  🔍 No markdown code blocks found, using raw response`);
      }
      
      this.log(`  📄 Final JSON length: ${responseText.length} characters`);
      this.log(`  📄 Final JSON preview: ${responseText.substring(0, 200)}...`);
      
      const result = JSON.parse(responseText);
      this.log(`  ✅ JSON parsed successfully`);
      this.log(`  📊 Found ${result.phrases ? result.phrases.length : 0} phrases`);
      
      if (result.phrases && result.phrases.length > 0) {
        this.log(`  📝 Phrases: ${result.phrases.map(p => p.text).join(', ')}`);
      }
      
      // Convert to our format
      return result.phrases.map(phrase => ({
        text: phrase.text,
        key: phrase.key,
        line: this.findLineNumber(content, phrase.text),
        context: this.findContext(content, phrase.text)
      }));

    } catch (error) {
      this.log(`  ❌ GPT analysis failed for ${filePath}: ${error.message}`);
      this.log(`  🔍 Error details: ${error.stack}`);
      return [];
    }
  }

  findLineNumber(content, text) {
    const lines = content.split('\n');
    for (let i = 0; i < lines.length; i++) {
      if (lines[i].includes(text)) {
        return i + 1;
      }
    }
    return 1;
  }

  findContext(content, text) {
    const lines = content.split('\n');
    for (let i = 0; i < lines.length; i++) {
      if (lines[i].includes(text)) {
        return lines[i].trim();
      }
    }
    return '';
  }
  
  async analyzeFile(filePath, content, strings) {
    const         analysis = {
          filePath: filePath,
          strings: strings,
          complexity: this.calculateComplexity(strings),
          needsUseTranslation: !content.includes('useTranslation'),
          hasI18nSetup: content.includes('next-i18next') || content.includes('react-i18next'),
          suggestedKeys: this.generateSuggestedKeys(strings),
          priority: this.calculatePriority(strings, content, filePath)
        };
    
    return analysis;
  }
  
  calculateComplexity(strings) {
    // Simple complexity calculation based on string count and types
    let complexity = strings.length;
    
    // Add complexity for different string types
    const typeWeights = {
      'jsx-text': 1,
      'button-text': 2,
      'notify-message': 3,
      'jsx-attr': 1,
      'link-text': 2,
      'label-text': 2
    };
    
    strings.forEach(string => {
      complexity += typeWeights[string.type] || 1;
    });
    
    return complexity;
  }
  
  generateSuggestedKeys(strings) {
    const keys = [];
    
    strings.forEach(string => {
      const key = this.stringToKey(string.text, string.type);
      keys.push({
        original: string.text,
        suggested: key,
        type: string.type
      });
    });
    
    return keys;
  }
  
  stringToKey(text, type) {
    // Convert string to semantic key
    let key = text.toLowerCase()
      .replace(/[^a-z0-9\s]/g, '') // Remove special characters
      .replace(/\s+/g, '_') // Replace spaces with underscores
      .substring(0, 30); // Limit length
    
    // Add type prefix
    const typePrefixes = {
      'button-text': 'button',
      'notify-message': 'notification',
      'link-text': 'link',
      'label-text': 'label',
      'jsx-text': 'text',
      'jsx-attr': 'attr'
    };
    
    const prefix = typePrefixes[type] || 'text';
    return `${prefix}.${key}`;
  }
  
  calculatePriority(strings, content, filePath) {
    // Calculate priority based on file importance and string count
    let priority = strings.length;
    
    // Higher priority for important files
    if (filePath.includes('Header') || filePath.includes('Menu')) priority += 10;
    if (filePath.includes('pages/')) priority += 5;
    if (filePath.includes('components/')) priority += 3;
    
    // Higher priority for files with many strings
    if (strings.length > 5) priority += 5;
    if (strings.length > 10) priority += 10;
    
    return priority;
  }
}

module.exports = AnalyzeTool;

File: pythonAgent/tools/analyze_tool.py
"""
Analyze tool for finding hardcoded strings in React files.
"""

import json
import re
from pathlib import Path
from typing import List, Dict, Any
from tools.base_tool import BaseTool


class AnalyzeTool(BaseTool):
    """Tool for analyzing React files to find hardcoded strings."""
    
    async def execute(self) -> dict:
        """Execute the analyze tool."""
        self.log('Analyzing files for hardcoded strings...')
        
        # Get files from agent's memory/state
        state = self.agent.state_manager.get_state()
        files = state.context.files_to_process or []
        
        self.log(f'  Analyzing {len(files)} files for hardcoded strings...')
        
        if len(files) == 0:
            self.log('  No files to analyze')
            return {'files_analyzed': 0, 'strings_found': 0, 'files': {}}
        
        analysis_results = {
            'files_analyzed': 0,
            'strings_found': 0,
            'files': {}
        }
        
        for file_path in files:
            if Path(file_path).exists():
                try:
                    content = Path(file_path).read_text(encoding='utf-8')
                    strings = await self._find_hardcoded_strings_with_claude(content, file_path)
                    
                    if len(strings) > 0:
                        analysis_results['files_analyzed'] += 1
                        analysis_results['strings_found'] += len(strings)
                        analysis_results['files'][file_path] = {
                            'strings': strings,
                            'count': len(strings),
                            'complexity': self._calculate_complexity(strings),
                            'needs_use_translation': 'useTranslation' not in content,
                            'has_i18n_setup': 'next-i18next' in content or 'react-i18next' in content
                        }
                        
                        self.log(f'  {file_path}: {len(strings)} strings found')
                    else:
                        self.log(f'  {file_path}: No strings found')
                        
                except Exception as error:
                    self.error(f'Error reading {file_path}: {error}')
            else:
                self.log(f'  File not found: {file_path}')
        
        # Store results in agent state and update phase
        self.agent.state_manager.update_context({
            'strings_found': analysis_results['strings_found'],
            'analysis_results': analysis_results
        })
        
        self.agent.state_manager.update_state({
            'phase': 'analyzing'
        })
        
        return analysis_results
    
    async def _find_hardcoded_strings_with_claude(self, content: str, file_path: str) -> List[Dict[str, Any]]:
        """Find hardcoded strings using Claude."""
        try:
            prompt = f"""Find ALL user-facing text in this React file. Return JSON with phrases array.

INCLUDE:
- Button text content
- Link text content  
- Label text content
- Alt text (alt="...")
- Placeholder text (placeholder="...")
- Aria labels (aria-label="...")
- Title attributes (title="...")
- Any visible text users see
- Error messages, notifications, tooltips
- Form labels and help text

EXCLUDE:
- HTML IDs, CSS classes, variable names
- Code logic, function names
- URLs, file paths, technical values
- Comments and console.log statements

Generate semantic keys like "button.save", "menu.my_books", "alt.logo", "placeholder.email"

Return only valid JSON: {{"phrases": [{{"text": "original", "key": "semantic.key"}}]}}

File: {file_path}

```jsx
{content}
```"""

            
            response = await self.ask_claude(prompt, 2000)
            
            # Extract JSON from markdown code blocks if present
            json_text = self._extract_json_from_response(response)
            
            try:
                result = json.loads(json_text)
            except json.JSONDecodeError as parse_error:
                self.log(f'  JSON parsing failed: {parse_error}')
                self.log(f'  Raw response: {json_text[:200]}...')
                return []
            
            phrase_count = len(result.get('phrases', []))
            
            if phrase_count > 0:
                phrases_text = ', '.join([p['text'] for p in result['phrases']])
                self.log(f'  Found {phrase_count} phrases: {phrases_text}')
            else:
                self.log('  No phrases found')
            
            # Convert to our format
            return [{
                'text': phrase['text'],
                'key': phrase['key'],
                'line': self._find_line_number(content, phrase['text']),
                'context': self._find_context(content, phrase['text'])
            } for phrase in result.get('phrases', [])]

        except Exception as error:
            self.error(f'Claude analysis failed for {file_path}: {error}')
            return []
    
    def _extract_json_from_response(self, response: str) -> str:
        """Extract JSON from Claude's response, handling markdown code blocks."""
        # Try multiple patterns to handle different response formats
        json_match = re.search(r'```(?:json)?\s*\n?(\{[\s\S]*?\})\s*```', response)
        if json_match:
            return json_match.group(1)
        
        # Try pattern without language specification
        json_match = re.search(r'```\s*\n?(\{[\s\S]*?\})\s*```', response)
        if json_match:
            return json_match.group(1)
        
        # Try pattern with just the opening brace
        json_match = re.search(r'(\{[\s\S]*?\})', response)
        if json_match:
            return json_match.group(1)
        
        return response
    
    def _find_line_number(self, content: str, text: str) -> int:
        """Find the line number where text appears."""
        lines = content.split('\n')
        for i, line in enumerate(lines):
            if text in line:
                return i + 1
        return 1
    
    def _find_context(self, content: str, text: str) -> str:
        """Find the context line where text appears."""
        lines = content.split('\n')
        for line in lines:
            if text in line:
                return line.strip()
        return ''
    
    def _calculate_complexity(self, strings: List[Dict[str, Any]]) -> int:
        """Calculate complexity based on string count and types."""
        complexity = len(strings)
        
        # Add complexity for different string types
        type_weights = {
            'jsx-text': 1,
            'button-text': 2,
            'notify-message': 3,
            'jsx-attr': 1,
            'link-text': 2,
            'label-text': 2
        }
        
        for string in strings:
            string_type = string.get('type', 'jsx-text')
            complexity += type_weights.get(string_type, 1)
        
        return complexity
